pkg/natdiscovery/remote_endpoint.go

```diff
@@ -0,0 +1,166 @@
+/*
+© 2021 Red Hat, Inc. and others.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package natdiscovery
+
+import (
+	"time"
+
+	"github.com/pkg/errors"
+
+	subv1 "github.com/submariner-io/submariner/pkg/apis/submariner.io/v1"
+	"github.com/submariner-io/submariner/pkg/types"
+)
+
+type endpointState int
+
+const (
+	testingPrivateAndPublicIPs endpointState = iota
+	waitingForResponse
+	selectedPublicIP
+	selectedPrivateIP
+	recheckTime                    = 2 * time.Second
+	totalTimeout                   = 60 * time.Second
+	publicToPrivateFailoverTimeout = 1 * time.Second
+)
+
+type remoteEndpointNAT struct {
+	endpoint               *types.SubmarinerEndpoint
+	state                  endpointState
+	lastCheck              time.Time
+	lastTransition         time.Time
+	started                time.Time
+	useNAT                 bool
+	useIP                  string
+	lastPublicIPRequestID  uint64
+	lastPrivateIPRequestID uint64
+}
+
+type NATEndpointInfo struct {
+	Endpoint subv1.EndpointSpec
+	UseNAT   bool
+	UseIP    string
+}
+
+func (rn *remoteEndpointNAT) GetNATEndpointInfo() *NATEndpointInfo {
+	return &NATEndpointInfo{
+		Endpoint: rn.endpoint.Spec,
+		UseNAT:   rn.useNAT,
+		UseIP:    rn.useIP,
+	}
+}
+
+func (rn *remoteEndpointNAT) Endpoint() *types.SubmarinerEndpoint {
+	// endpoint is not mutable, so no locking in this case
+	return rn.endpoint
+}
+
+func (rn *remoteEndpointNAT) UseIP() string {
+	return rn.useIP
+}
+
+func (rn *remoteEndpointNAT) UseNAT() bool {
+	return rn.useNAT
+}
```

Comment:
We don't actually need these methods since we don't export it and callers can access the fields directly anyway.
 
```suggestion
```

Reply:
correct.

Reply:
done.